File: leetcode3/남효정/4006. Count Valid Prefixes.py
class Solution:
    def countValidPrefixes(self, s: str) -> int:
        ans, diff = 0, 0

        # 0, 1의 개수가 서로 1 이하면 됨
        for char in s:
            if char == '1':
                diff += 1
            else:
                diff -= 1
            
            if abs(diff) <= 1:
                ans += 1

        return ans

File: leetcode3/염혜정/4006. Count Valid Prefixes.java
// prefix의 길이가 짝수면 0과 1의 개수가 같아야 하고
// prefix의 길이가 홀수면 0과 1의 개수의 차이가 1이어야 함
// O(n)

class Solution {
    public int countValidPrefixes(String s) {
        int[] cnt = new int[2];
        int result = 0;

        for (int i = 0; i<s.length(); i++) {
            cnt[s.charAt(i) - '0']++;
            
            if ((i+1) % 2 == 0) {
                if (cnt[0] == cnt[1]) result++;
            } else {
                if (Math.abs(cnt[0] - cnt[1]) == 1) result++;
            }
        }
        return result;
    }
}
